af/shovel/centrifugation.py

```diff
@@ -326,163 +329,491 @@ def __flush_stubborn(self):
-    columns = ('filename', 'bucket_date', 'code_ver', 'file_size', 'file_crc32', 'file_sha1')
-    fmt = '{}\t' + bucket + '\t' + str(CODE_VER) + '\t{:d}\t{:d}\t\\\\x{}\n'
-    with closing(PGCopyFrom(pgconn, table, columns=columns)) as wbuf, \
-        gzip.GzipFile(autoclaved_index, 'r') as indexfd:
-        filename = None
-        for _, doc in autoclaving.stream_json_blobs(indexfd):
-            if 'file"' in doc:
-                doc = ujson.loads(doc)
-                if doc['type'] == 'file':
-                    if filename is not None:
-                        raise RuntimeError('Corrupt index file', autoclaved_index, doc)
-                    filename = doc['filename']
-                elif doc['type'] == '/file':
-                    if filename is None:
-                        raise RuntimeError('Corrupt index file', autoclaved_index, doc)
-                    doc['filename'], filename = filename, None
-                    wbuf.write(fmt.format(pg_quote(doc['filename']), doc['file_size'],
-                            doc['file_crc32'], b64decode(doc['file_sha1']).encode('hex')))
-
-def copy_reports_from_index(pgconn, autoclaved_index, table, autoclaved_no):
-    # autoclaved/${bucket}/index.json.gz | filter "type=report" > report_meta.table
-    columns = ('autoclaved_no', 'badtail', 'textname', 'orig_sha1')
-    with closing(PGCopyFrom(pgconn, table, columns=columns)) as wbuf, \
-        gzip.GzipFile(autoclaved_index, 'r') as indexfd:
-        rowno = None
-        for _, doc in autoclaving.stream_json_blobs(indexfd):
-            if '"datum"' not in doc: # till someone submits report with {"textname":"datum"}...
-                doc = ujson.loads(doc)
-                t = doc['type']
-                if t == 'file':
-                    rowno = autoclaved_no[doc['filename']]
-                elif t == 'report':
-                    report = doc
-                elif t == '/report' and report['textname'] not in DUPLICATE_REPORTS:
-                    wbuf.write('{:d}\t{}\t{}\t\\\\x{}\n'.format(rowno, pg_quote(doc.get('src_cutoff')),
-                            pg_quote(report['textname']), b64decode(report['orig_sha1']).encode('hex')))
-
-def copy_measurements_from_index(pgconn, autoclaved_index, table, report_no):
-    # autoclaved/${bucket}/index.json.gz | filter "type=datum" > measurement_meta.table
-    columns = ('report_no', 'frame_off', 'frame_size', 'intra_off', 'intra_size', 'orig_sha1')
-    with closing(PGCopyFrom(pgconn, table, columns=columns)) as wbuf, \
-        gzip.GzipFile(autoclaved_index, 'r') as indexfd:
-        rowno = None
-        # Separate `use_report` variable is used instead of `rowno=None` to
-        # enforce implicit file format checking.
-        use_report = True
+            raise AssertionError('BUG: should never happen')
+        dest_wbuf[(doc_type, key)].write(copy_row)
+    with gzip.GzipFile(autoclaved_index, 'r') as indexfd:
```

Comment:
This whole function is very long, maybe  it could be refactored into a smaller function (the switch case inside of the for loop) for readability.

Reply:
ACK